presets/inference/vllm/inference_api.py

```diff
@@ -58,15 +63,55 @@ def find_max_available_seq_len(engine_config: EngineConfig) -> int:
         observability_config=engine_config.observability_config,
     )
 
-    # see https://github.com/vllm-project/vllm/blob/v0.6.3/vllm/engine/llm_engine.py#L477
-    num_gpu_blocks, _ = executor.determine_num_available_blocks()
+    max_probe_steps = 6
+    if os.getenv("MAX_PROBE_STEPS") is not None:
+        try:
+            max_probe_steps = int(os.getenv("MAX_PROBE_STEPS"))
+        except ValueError:
+            raise ValueError("MAX_PROBE_STEPS must be an integer.")
+
+    model_max_blocks = int(engine_config.model_config.max_model_len / engine_config.cache_config.block_size)
+    res = binary_search_with_limited_steps(model_max_blocks, max_probe_steps, lambda x: is_context_length_safe(executor, x))
 
     # release memory
     del executor
     gc.collect()
     torch.cuda.empty_cache()
 
-    return engine_config.cache_config.block_size * num_gpu_blocks
+    return engine_config.cache_config.block_size * res
+
+def binary_search_with_limited_steps(upper: int, max_probe_steps: int, is_valid_fn: Callable[[int], bool]) -> int:
```

Comment:
```
def find_max_valid_context_length(max_context_length: int, max_probe_steps: int, is_valid_fn: Callable[[int], bool]) -> int:
    """
    Finds the maximum valid context length for inference without causing an out-of-memory (OOM) error.
    
    Parameters:
    - max_context_length (int): The initial upper bound for the context length to be tested.
    - max_probe_steps (int): Maximum number of steps to probe for a valid context length.
    - is_valid_fn (Callable[[int], bool]): A function that checks if a given context length is valid.
    
    Returns:
    - int: The maximum valid context length that does not cause an OOM error.
    """
    low, high, probe_steps = 0, max_context_length * 2, 0
    
    while low < high and probe_steps < max_probe_steps:
        mid = (low + high + 1) // 2
        if mid > max_context_length:
            break

        if is_valid_fn(mid):
            low = mid  # Move low pointer up if valid
        else:
            high = mid - 1  # Move high pointer down if not valid

        probe_steps += 1

    return low
```

Reply:
your suggestion is good. but I don't want to introduce any context info about the invoker for this function. It's just a pure helper function. The parameters passed in are well-documented.
We can add a comment about this algorithm.